File: mac/ClipSync/UpdateNotificationManager.swift
import Foundation
import UserNotifications

// Purpose: UpdateNotificationManager handles pending update storage and notification display.
// Responsibilities: Manages update notification state for dialog display.
// Usage: Save pending updates when FCM notification is received, retrieve in HomeScreen.
class UpdateNotificationManager {
    static let shared = UpdateNotificationManager()
    
    private let defaults = UserDefaults.standard
    private let KEY_PENDING_VERSION = "pending_update_version"
    private let KEY_PENDING_URL = "pending_update_url"
    private let KEY_PENDING_NOTES = "pending_update_notes"
    private let KEY_HAS_PENDING = "has_pending_update"
    
    private init() {}
    
    
    // Purpose: Saves pending update information.
    // Parameters: version, downloadUrl, releaseNotes.
    // Returns: Void unless returned explicitly.
    // Notes: Data persists until cleared by user.
    func savePendingUpdate(version: String, downloadUrl: String, releaseNotes: String) {
        defaults.set(version, forKey: KEY_PENDING_VERSION)
        defaults.set(downloadUrl, forKey: KEY_PENDING_URL)
        defaults.set(releaseNotes, forKey: KEY_PENDING_NOTES)
        defaults.set(true, forKey: KEY_HAS_PENDING)
        defaults.synchronize()
        
        print("✅ Pending update saved: \(version)")
    }
    
    
    // Purpose: Retrieves pending update information if available.
    // Parameters: No parameters.
    // Returns: UpdateInfo? - update details or nil.
    // Notes: Returns nil if no pending update.
    func getPendingUpdate() -> UpdateInfo? {
        guard defaults.bool(forKey: KEY_HAS_PENDING),
              let version = defaults.string(forKey: KEY_PENDING_VERSION),
              let url = defaults.string(forKey: KEY_PENDING_URL) else {
            return nil
        }
        
        let notes = defaults.string(forKey: KEY_PENDING_NOTES) ?? "New update available!"
        return UpdateInfo(version: version, downloadUrl: url, releaseNotes: notes)
    }
    
    
    // Purpose: Clears pending update information.
    // Parameters: No parameters.
    // Returns: Void unless returned explicitly.
    // Notes: Call after user dismisses update dialog.
    func clearPendingUpdate() {
        defaults.removeObject(forKey: KEY_PENDING_VERSION)
        defaults.removeObject(forKey: KEY_PENDING_URL)
        defaults.removeObject(forKey: KEY_PENDING_NOTES)
        defaults.set(false, forKey: KEY_HAS_PENDING)
        defaults.synchronize()
    }
    
    
    // Purpose: Data structure for update information.
    // Responsibilities: Holds update metadata.
    // Usage: Return type for getPendingUpdate().
    struct UpdateInfo {
        let version: String
        let downloadUrl: String
        let releaseNotes: String
    }
}

File: android/app/src/main/java/com/bunty/clipsync/UpdateNotificationManager.kt
package com.bunty.clipsync

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat

/**
 * UpdateNotificationManager handles update notifications and pending update storage.
 */
object UpdateNotificationManager {

    private const val CHANNEL_ID = "clipsync_updates"
    private const val CHANNEL_NAME = "App Updates"
    private const val NOTIFICATION_ID = 1001

    private const val PREFS_NAME = "update_prefs"
    private const val KEY_PENDING_VERSION = "pending_version"
    private const val KEY_PENDING_URL = "pending_url"
    private const val KEY_PENDING_NOTES = "pending_notes"
    private const val KEY_HAS_PENDING = "has_pending_update"

    /**
     * Shows an update notification when a new version is available.
     */
    fun showUpdateNotification(context: Context, version: String, releaseNotes: String) {
        createNotificationChannel(context)

        val intent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            putExtra("show_update_dialog", true)
        }

        val pendingIntent = PendingIntent.getActivity(
            context,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_launcher_foreground)
            .setContentTitle("ClipSync Update Available 🚀")
            .setContentText("Version $version is now available!")
            .setStyle(
                NotificationCompat.BigTextStyle()
                    .bigText("Version $version is now available!\n\n${releaseNotes.take(100)}")
            )
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setAutoCancel(true)
            .setContentIntent(pendingIntent)
            .build()

        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.notify(NOTIFICATION_ID, notification)
    }

    /**
     * Creates notification channel for update notifications (Android 8.0+).
     */
    private fun createNotificationChannel(context: Context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_HIGH
            ).apply {
                description = "Notifications for app updates"
            }

            val notificationManager = context.getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }
    }

    /**
     * Saves pending update information for later display.
     */
    fun savePendingUpdate(context: Context, version: String, downloadUrl: String, releaseNotes: String) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit().apply {
            putString(KEY_PENDING_VERSION, version)
            putString(KEY_PENDING_URL, downloadUrl)
            putString(KEY_PENDING_NOTES, releaseNotes)
            putBoolean(KEY_HAS_PENDING, true)
            apply()
        }
    }

    /**
     * Returns pending update info if available.
     */
    fun getPendingUpdate(context: Context): UpdateInfo? {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val hasPending = prefs.getBoolean(KEY_HAS_PENDING, false)

        if (!hasPending) return null

        val version = prefs.getString(KEY_PENDING_VERSION, null) ?: return null
        val url = prefs.getString(KEY_PENDING_URL, null) ?: return null
        val notes = prefs.getString(KEY_PENDING_NOTES, "New update available!") ?: "New update available!"

        return UpdateInfo(version, url, notes)
    }

    /**
     * Clears pending update information.
     */
    fun clearPendingUpdate(context: Context) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit().apply {
            remove(KEY_PENDING_VERSION)
            remove(KEY_PENDING_URL)
            remove(KEY_PENDING_NOTES)
            putBoolean(KEY_HAS_PENDING, false)
            apply()
        }
    }

    /**
     * Data class for update information.
     */
    data class UpdateInfo(
        val version: String,
        val downloadUrl: String,
        val releaseNotes: String
    )
}
